subsys/net/ip/tcp2.c

```diff
@@ -1070,6 +1075,9 @@ static void tcp_resend_data(struct k_work *work)
 
 			goto out;
 		}
+	} else if (ret == -ENODATA) {
+		conn->data_mode = TCP_DATA_MODE_SEND;
```

Comment:
I just wonder if this is the right place to change the data_mode, or if this rather should be done in the TCP state machine.  This would be achieved if we removed the check at https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp2.c#L1840, which as I wrote earlier I don't think is correct. @jukkar I think you have the best insight into the TCP2 codebase, do you have an opinion on this?

Reply:
That is a good question and I do not have clear answer to this. Tweaking/removing the check on line 1840 might be a good start for a fix. Sorry, but is difficult to give direct answer as there might be unseen consequences when changing the TCP code. So I would suggest we experiment with your suggestion before accepting the changes.

Reply:
With https://github.com/zephyrproject-rtos/zephyr/commit/89212a7fbf5fcd8e4d661c016344ae4bf2d46f53 applying this patch, my test described in #32564 did not crash system and the test ran for 240 hours. If this patch is not applied, the NULL pointer access may occur with CONFIG_ASSERT=n.

Reply:
Ok, I think I can accept the fix as it is then. given that you've put it through so severe testing and it actually helps with the problem.